### PR TITLE
Updated GNC Utility Functions and Attitude Estimator

### LINKED ISSUE #159: Update CXX Attitude Estimator to Work Assuming Perfect Sensors
If we want to get some version of flight software up and running in a HOOTL environment with the new attitude controller, we'll need to make some small changes to the attitude estimator interface. Most notably, the finite difference method to determine `w_body` should be removed because we will be relying on filtered gyroscope measurements in the future.

Obviously, this implementation will be removed in the future.

### LINKED ISSUE #161: Refactor the Triad Utility Function
Currently, the arguments to the triad utility function are named assuming we are only using it for attitude determination given sun and magnetic field vectors. This isn't actually the case, so it'd be nice if we could refactor it to take more general arguments.

### LINKED ISSUE #163: Improve NaN Support in CXX GNC Utility Functions
All of the utility functions don't necessarily conform to setting their outputs to NaN in the case of a failure. We need to do a pass through the utility functions and write tests to check for this.

### PR DESCRIPTION
**Pull request by kylekrol**
# Updated GNC Utility Functions and Attitude Estimator

Fixes #159.
Fixes #161.
Fixes #163.

### Summary of changes
- Put in an explicit threshold on how "close" the triad vector pairs can be before an NaNs quaternion is returned (currently set to `1.0` degrees).
- Add asserts to enforce invariants on GNC utility functions.
- Update GNC utility function documentation to comment on NaN handling.
- Add in `w_sat` as an input to the orbit estimator and remove the finite differencing of the attitude quaternion. We currently are just passing the measured omega through as the "estimate", however, this will get replaced by a full filter in the future.

### Ptest Effects

Not applicable at the moment. Although, with this merged, work can commence on [this](https://github.com/pathfinder-for-autonomous-navigation/FlightSoftware/issues/287).

### Testing
Deferring testing for [mex functions](https://github.com/pathfinder-for-autonomous-navigation/psim/issues/160) and [fsw integration](https://github.com/pathfinder-for-autonomous-navigation/FlightSoftware/issues/287).

### Constants

No change.

### Documentation Evidence

See inline documentation.

### CODE REVIEW

File: include/gnc/utilities.hpp [hunk 1]
```diff
@@ -15,60 +20,93 @@ namespace gnc {
...
- * 
- *  REQUIRES: N_* and B_* vectors must be unit vectors. */
+ *  @param[in]  R1 First known vector in a known frame.
+ *  @param[in]  R2 Second known vector in a known frame.
+ *  @param[in]  r1 First measurement vector in an unknown frame.
+ *  @param[in]  r2 Second measurement vector in an unknown frame.
+ *  @param[out] q  Quaternion transforming from the unknown to known frame.
+ *  Given a set of unit vector from a known frame and then unknown frame, this
+ *  function calculates the rotation required to transform from the latter to
+ *  the former. If R1 and R2, or r1 and r2 are within a degree of one another,
```
Comment (nhz2) [inline include/gnc/utilities.hpp:88]: I would add that q will rotate the first vector exactly, and the second vector as best it can.

Comment (kylekrol) [reply include/gnc/utilities.hpp:88]: Oh shoot yep - will do

Comment (nhz2) [reply include/gnc/utilities.hpp:88]: Also that q is quat_unknown_known so it should rotate from known frame to unknown frame. Also, there are no MATLAB tests for triad, its copied from Justin's Simulink, and the docs in `utl_triad.m` aren't very helpful. #167

File: src/gnc_attitude_estimator.cpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+static void estimate_q_body_eci(double t, lin::Vector3d const &r_ecef,
+    lin::Vector3f const &s_body, lin::Vector3f const &b_body,
+    lin::Vector4f &q_body_eci) {
+  lin::Vector3f s_eci, b_eci;
+
+  // Default to NaNs
+  q_body_eci = lin::nans<lin::Vector4f>();
+
+  // Determine the expected magnetic field
+  env::magnetic_field(t, lin::Vector3f(r_ecef), b_eci);
```
Comment (nhz2) [inline src/gnc_attitude_estimator.cpp:38]: I think this should be b_ecef

File: include/gnc/utilities.hpp (hunk 1 above)
Comment (kylekrol) [reply include/gnc/utilities.hpp:88]: Oh wait really? Welp, I may need to rework some documentation in other places as well. Really not sure where I got that from. Thanks for keeping an eye out!

**Review by nhz2 — CHANGES_REQUESTED** [review]
The magnetic field in ECEF should be rotated into the ECI frame for triad.

File: src/gnc_attitude_estimator.cpp (hunk 2 above)
Comment (kylekrol) [reply src/gnc_attitude_estimator.cpp:38]: Oh my yes indeed.

Comment (kylekrol) [reply src/gnc_attitude_estimator.cpp:38]: Alright this should be fixed.

File: include/gnc/utilities.hpp (hunk 1 above)
Comment (kylekrol) [reply include/gnc/utilities.hpp:88]: And greatly improved the documentation in CXX at least for triad.

**Comment by kylekrol** [comment]
Oh shoot I'm sorry - this is the wrong branch

**Pull request closed without merging**